# **F-test**

In [32]:
import pandas as pd
import numpy as np

from statsmodels.formula.api import ols

from scipy.stats import f # # import F-distribution and critical values

In [33]:
# import the dataset
pancake_yummy = pd.read_csv('sleep75.csv')

In [34]:
pancake_yummy.head(5)

,age,black,case,clerical,construc,educ,earns74,gdhlth,inlf,leis1,...,spwrk75,totwrk,union,worknrm,workscnd,exper,yngkid,yrsmarr,hrwage,agesq
0,32,0,1,0.0,0.0,12,0,0,1,3529,...,0,3438,0,3438,0,14,0,13,7.070004,1024
1,31,0,2,0.0,0.0,14,9500,1,1,2140,...,0,5020,0,5020,0,11,0,0,1.429999,961
2,44,0,3,0.0,0.0,17,42500,1,1,4595,...,1,2815,0,2815,0,21,0,0,20.529997,1936
3,30,0,4,0.0,0.0,12,42500,1,1,3211,...,1,3786,0,3786,0,12,0,12,9.619998,900
4,64,0,5,0.0,0.0,14,2500,1,1,4052,...,1,2580,0,2580,0,44,0,33,2.750000,4096


In [35]:
pancake_yummy.shape

(706, 34)

In [36]:
pancake_yummy.columns.tolist()

['age',
 'black',
 'case',
 'clerical',
 'construc',
 'educ',
 'earns74',
 'gdhlth',
 'inlf',
 'leis1',
 'leis2',
 'leis3',
 'smsa',
 'lhrwage',
 'lothinc',
 'male',
 'marr',
 'prot',
 'rlxall',
 'selfe',
 'sleep',
 'slpnaps',
 'south',
 'spsepay',
 'spwrk75',
 'totwrk',
 'union',
 'worknrm',
 'workscnd',
 'exper',
 'yngkid',
 'yrsmarr',
 'hrwage',
 'agesq']

## Estimate the regression

### Non-robust vs Robust

A robust test in econometrics usually means conducting standard hypothesis tests using variance estimators that remain consistent under heteroskedasticity, serial correlation, or clustering. 

Robust test differs from a non-robust test in the command '.fit()'. Here we will be fitting a model with a robust HC3 estimate of the covariance matrix.

To carry out a robust test add ".fit(cov_type='HC3')".

The command ".fit(cov_type='HC3')" implies computing heteroskedasticity-robust standard errors using the HC3 covariance estimator.

Doing a robust or a non-robust test is taken account when evaluating a regression. The steps after it are communicated this adjustment from estimating the regression.

In [37]:
# Non-robust
pancakes_should_be_eaten_model_1 = ols(formula='sleep~totwrk+age+np.square(age)+south+male+smsa', data=pancake_yummy).fit()

In [38]:
# Robust
cupcakes_should_be_eaten_model_2 = ols(formula='sleep~totwrk+age+np.square(age)+south+male+smsa', data=pancake_yummy).fit(cov_type='HC3')

## Significance of the entire regression

#### Get the F-statistic value

In [39]:
# Use this function to get the F-statistic value for the whole model
# model.fvalue
pancakes_should_be_eaten_model_1.fvalue.round(3)

17.615

#### Calculate the critical value

In [40]:
# significance_level (α)= 0.05 = 5%
# dfd = degrees_of_freedom = n-k-1 = 706-7-1 (for this specific case)
# n - number of observations
# dfn = k - number of regressors in the model (number of X's)
# f.isf(q=α, dfn=k, dfd=n-k-1)
f.isf(q=0.05,dfn=6,dfd=706-6-1).round(3)

2.112

#### Compare the F-statistic value with the critical value

If |F_stat| < F_cr => we fail to reject the null hypothesis (H0) => the regression is not statistically significant (at α significance level). The B coefficients can be taken equal to 0.

If |F_stat| > F_cr => we reject the null hypothesis (H0) => the regression is statistically significant (at α significance level). At least one of the B coefficients canbot be taken equal to 0.

#### OR compare the P-value with the significance level

If the p-value > alpha/α/significance level => we fail to reject the H0.

If the p-value < alpha/α/significance level => we to reject the H0.

In [41]:
# Use this function to get the P-value for the whole model
# model.f_pvalue
pancakes_should_be_eaten_model_1.f_pvalue

4.680414780965923e-19

#### Example

F-statistic VS Critical value<br>
17.615 > 2.112 => the regression is statistically significant (at 5% significance level).

P-value VS Significance level<br>
4.680414780965923e-19 < 0.05 => the regression is statistically significant (at 5% significance level).

## Significance of several coefficients

With the F-test we can estimate the statistical significance of several coefficents at once, not necessarily the whole model.

#### Get the F-statistic value

In [42]:
# To run the F-test for several regressors use the following function
# model.f_test(X1=X2=0)
# OR
# model.f_test(X1=0, X2=0)
# Where substitute X1, X2, ... with the regressors from the model
# If you have np.square(X) or np.log(X), include them with np. functions
pancakes_should_be_eaten_model_1.f_test('totwrk=age=0')

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=44.68306084703249, p=5.474611374876762e-19, df_denom=699, df_num=2>

The results of the F-test for several coefficients show
- the F-statistic value for the coefficients selected from the model
- the P-value
- the degrees of freedom

#### Calculate the critical value

While determining statistical signifficance of several coefficients you need to calculate the critical value with the parameters **df_denom** and **df_num** from the F-test results.

In [43]:
f.isf(q=0.05, dfn=2,dfd = 699).round(2)

3.01

#### Compare the F-statistic value with the critical value

#### OR compare the P-value with the significance level

The principles and the interpretation are identical to those when determining significance of the entire regression.